power_grid_model_io_native_c/power_grid_model_io_native/include/power_grid_model_io_native/pgm_vnf_converter/pgm_vnf_converter.hpp

```diff
@@ -83,123 +79,53 @@ inline PgmVnfConverter::PgmVnfConverter(std::string_view buffer, ExperimentalFea
+    this->deserialized_data_.set_construction_complete();
     constexpr auto const& meta_data = pgm::meta_data::meta_data_gen::meta_data;
-    pgm::ConstDataset const const_dataset = make_const_dataset(nodes, meta_data);
-
+    pgm::ConstDataset const const_dataset = make_const_dataset(meta_data);
     std::string const serialized_pgm_data = serialize_data(const_dataset);
-
-    // 1. our vnf importer it directly understands the vnf format
-    // 2. convert vnf like dataset to internal types (take raw data and convert it to pgm component container)
-    // 3. another function which 1. makes const dataset view from pgm component container  2. const dataset to
-    // serializer
-
     this->serialized_data_ = serialized_pgm_data;
 }
 
 inline void PgmVnfConverter::set_file_buffer(std::string_view file_buffer) { this->buffer_ = file_buffer; }
 
-inline void PgmVnfConverter::set_deserialized_dataset(pgm::WritableDataset* data) { this->deserialized_data_ = data; }
+inline void PgmVnfConverter::set_deserialized_dataset(PgmInput data) { this->deserialized_data_ = std::move(data); }
 
 inline std::string_view PgmVnfConverter::get_file_buffer() const { return this->buffer_; }
 
-inline pgm::WritableDataset* PgmVnfConverter::get_deserialized_dataset() { return this->deserialized_data_; }
+inline PgmInput PgmVnfConverter::get_deserialized_dataset() const { return this->deserialized_data_; }
 
 inline std::string const& PgmVnfConverter::get_serialized_data() const { return this->serialized_data_; }
 
-inline std::vector<pgm::NodeInput> PgmVnfConverter::convert_node_input() {
-    auto node_inputs = this->parsed_nodes_;
-    // std::vector<pgm::Node> nodes;
+inline void PgmVnfConverter::convert_node_input() {
+    std::vector<pgm::NodeInput> nodes;
 
-    // nodes.reserve(node_inputs.size());
-    // for (auto const& node_input : node_inputs) {
-    //     nodes.emplace_back(node_input);
-    // }
+    for (auto const& node : this->parsed_vnf_data_.iter<VnfNode>()) {
+        // Lookup PGM node id and get vnf node u_nom value
+        auto node_id = id_lookup_[node.guid];
+        auto vnfnode_unom = node.u_nom;
 
-    return node_inputs;
-}
+        // add u_nom multiplier when known
+        this->deserialized_data_.emplace<pgm::NodeInput>(node_id, node_id, vnfnode_unom);
 
-inline void PgmVnfConverter::convert_line_input() {
-    // Implementation
-}
-
-inline void PgmVnfConverter::convert_sources_input() {
-    // Implementation
-}
-
-inline void PgmVnfConverter::convert_sym_loads_input() {
-    // Implementation
-}
-
-inline void PgmVnfConverter::convert_shunts_input() {
-    // Implementation
-}
-
-inline void PgmVnfConverter::convert_transformer_input() {
-    // Implementation
-}
-
-inline void PgmVnfConverter::convert_sym_gens_input() {
-    // Implementation
-}
-
-inline void PgmVnfConverter::convert_links_input() {
-    // Implementation
+        nodes.emplace_back(pgm::NodeInput{node_id, vnfnode_unom});
+    }
+    this->nodes_ = nodes;
```

Comment:
either prevent copy by using `move`
```suggestion
    this->nodes_ = std::move(nodes);
```
or use `this->nodes_` from the get-go in this function (either directly or by using `auto& nodes = this->nodes_;`)